### ISSUE TITLE
Does T5 supports gradient accumulation?

### ISSUE THREAD

**Issue by MGithubGA** [google-research/text-to-text-transfer-transformer#687]
I am training a T5-small using GPUs. However I got an OOM when I set tokens_per_batch as 1048576.

I am wondering whehter T5 supports gradient accumulation. 

Thanks! @craffel

**Comment by craffel** [comment]
Yes, it does. https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/utils.py#L2208

**Comment by MGithubGA** [comment]
Thank @craffel  for your reply!

When I try to set micro batch size by `--gin_param="serialize_num_microbatches.tokens_per_microbatch_per_replica = 2048"`, I face anothor issue:

Under the same tokens_per_microbatch_per_replica, a larger tokens_per_batch leads to OOM while a lower tokens_per_batch works well. 

For example, I can pretrain T5-small with tokens_per_batch=32768 and tokens_per_microbatch_per_replica=2048 on 8 32G V100 GPUs. However, if I use tokens_per_batch=1048576, I have to set tokens_per_microbatch_per_replica=512 and mesh_shape = 'model:2,batch:4' to prevent OOM.

Generally, gradient accumulation does not require such large extra GPU memory. Could you please help me with the problem?

Thanks!

**Comment by VictorSanh** [comment]
> Thank @craffel for your reply!
> 
> When I try to set micro batch size by `--gin_param="serialize_num_microbatches.tokens_per_microbatch_per_replica = 2048"`, I face anothor issue:
> 
> Under the same tokens_per_microbatch_per_replica, a larger tokens_per_batch leads to OOM while a lower tokens_per_batch works well.
> 
> For example, I can pretrain T5-small with tokens_per_batch=32768 and tokens_per_microbatch_per_replica=2048 on 8 32G V100 GPUs. However, if I use tokens_per_batch=1048576, I have to set tokens_per_microbatch_per_replica=512 and mesh_shape = 'model:2,batch:4' to prevent OOM.
> 
> Generally, gradient accumulation does not require such large extra GPU memory. Could you please help me with the problem?
> 
> Thanks!

@MGithubGA I am running into the same problem: fixing `serialize_num_microbatches.tokens_per_microbatch_per_replica` while increasing `tokens_per_batch` eventually leads to OOM issues (on GPU). Any hints on why this is happening?

ping @craffel

**Comment by craffel** [comment]
I was able to get around this by downgrading TensorFlow to 2.2.1. I think there was a change in how XLA decides to do rematerialization that results in OOMs.
